_posts/2019-02-11-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+C++ for different platforms. Or maybe you maintain a popular JavaScript library
+that users only want parts of to keep their .js files small.
+
+These are examples of *configuration*: the process of building the same *code*
+with different *settings* to customize it for specific needs. In Bazel-speak,
+this means building a
+[target](https://docs.bazel.build/versions/master/build-ref.html#targets) with
+a set of
+[flags](https://docs.bazel.build/versions/master/user-manual.html#flags-options):
+
+```sh
+$ bazel build //my:cc_binary --cpu=x86
+$ bazel build //my:cc_binary --cpu=arm
+$ bazel build //my:android_binary --android_sdk=@androidsdk//:sdk-28
+```
+
+Bazel's *configurability work* is an ongoing effort to make these tasks simple,
+flexible, and powerful. 
+
+## The Problem
+
+Bazel is a
+[powerful](https://docs.bazel.build/versions/master/bazel-vision.html) build
+tool that's especially suited for large codebases with
+[multiple languages](https://blog.bazel.build/2018/12/05/multilanguage-build-system.html).
+But it grew out of Google, which historically wrote code for fleets of identical
+Linux servers with little need for customization. What customization was needed
+was achieved with
+[ad hoc flags and logic](https://source.bazel.build/bazel/+/144912e7b7a86b45e07f79e76f6fed20890acb36:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=258)
+built straight into the tool.
+
+
+This is no longer true inside or outside of Google. Modern software targets
+phones, cloud, servers, desktops, [smart
+devices](https://en.wikipedia.org/wiki/Smart_device) and more, and must offer
+increasingly flexible support for any combination of these platforms.
+
+This means
+
+    $ bazel build //my:binary
+
+isn't enough to describe what you want to do. What platform(s) are you
+targeting? What features do you want to include? What if you want to build a
+complex app with generated sources, client and server modules, native
+extensions, and test data?
+
+Bazel's historical approach of ad hoc flags:
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+is lacking in many ways:
+
+1. `--cpu` only accepts values explicitly supported by C++ rules. Even if you're
+   not building C++.
+1. `--crosstool_top`, which specifies exactly what command line should compile
+   your code, is even more tied to C++. Other languages may be organized
+   completely differently.
+1. `--define` is completely unstructured and unreadable in Starlark.
+1. `$ bazel build //my:all` sets the same flags for all targets, even if
+   different targets need different flags.
+1. `//my:binary`'s deps must use the same settings as their parent.
+1. It might be hard to remember which flags are needed by which targets,
+   especially across users and dev/test/prod machines. This makes it hard to
+   be confident in the integrity of your builds.
+
+Another approach might be to set flags directly in targets that need them, let
```

Comment:
The structure of this paragraph could be improved just a little bit, because it buries the lede. What about:

> There is a better way to solve the platforms problem, and it's the one we've decided to take:
> - Set flags on a per-target level
> - Give rule owners a system to design and consume flags
> - Provide standard APIs for cross-language concepts like *platform* and *cpu*
> 
> This makes configuration more flexible and tied to individual targets, rather than the system as a whole and will let developers configure projects on their own terms.

There might be factual issues in what I've written above, but what do you think about that organization for this paragraph? (I'm only picking on it because I think it's an important one.)

Reply:
Okay, so I somewhat integrated this suggestion. I still kept a lot of the original phrasing but moved the intention & benefit straight to the top. We can talk more about the phrasing if you like.

The only reason I didn't split out into bullet points is to keep it as concise as possible. My rough heuristic I've been applying throughout is that each section is fully viewable on my 13" MacBook screen. I left more leeway for the meaty sections. But, just as you expressed earlier, this background / context section shouldn't get eye-rollingly long.

I'm happy to bullet point it if you think that doesn't matter here and reads better given the importance of the paragraph.

Reply:
I really like the new version.